xcodecommand/xcpretty.go

```diff
@@ -0,0 +1,91 @@
+package xcodecommand
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"os"
+	"os/exec"
+
+	"github.com/bitrise-io/go-utils/v2/command"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+type xcprettyCommandRunner struct {
+	logger         log.Logger
+	commandFactory command.Factory
+}
+
+func NewXcprettyCommandRunner(logger log.Logger, commandFactory command.Factory) Runner {
+	return &xcprettyCommandRunner{
+		logger:         logger,
+		commandFactory: commandFactory,
+	}
+}
+
+func (c *xcprettyCommandRunner) Run(workDir string, xcodebuildArgs []string, xcprettyArgs []string) (Output, error) {
+	var (
+		buildOutBuffer         bytes.Buffer
+		pipeReader, pipeWriter = io.Pipe()
+		buildOutWriter         = io.MultiWriter(&buildOutBuffer, pipeWriter)
+		prettyOutWriter        = os.Stdout
+	)
+
+	buildCmd := c.commandFactory.Create("xcodebuild", xcodebuildArgs, &command.Opts{
+		Stdout: buildOutWriter,
+		Stderr: buildOutWriter,
+		Env:    xcodeCommandEnvs,
+		Dir:    workDir,
+	})
+
+	prettyCmd := c.commandFactory.Create("xcpretty", xcprettyArgs, &command.Opts{
+		Stdin:  pipeReader,
+		Stdout: prettyOutWriter,
+		Stderr: prettyOutWriter,
+	})
+
+	c.logger.TPrintf("$ set -o pipefail && %s | %v", buildCmd.PrintableCommandArgs(), prettyCmd.PrintableCommandArgs())
+
+	if err := buildCmd.Start(); err != nil {
+		return Output{
+			RawOut:   buildOutBuffer.Bytes(),
+			ExitCode: 1,
+		}, err
+	}
+	if err := prettyCmd.Start(); err != nil {
+		return Output{
+			RawOut:   buildOutBuffer.Bytes(),
+			ExitCode: 1,
+		}, err
+	}
+
+	defer func() {
```

Comment:
This should be higher in the function. At this point there are two return statements above this.

Reply:
Moved pipe close after declaration. Wait is only needs to be called if Start() is successful, left that in place.